Source/DafnyRuntime/DafnyRuntimeRust/src/lib.rs

```diff
@@ -9,6 +16,346 @@ pub use num::NumCast;
+    use std::collections::HashMap;
+    use std::rc::Rc;
+    use std::hash::Hash;
+    pub fn dafny_int_to_bigint(i: &Rc<BigInt>) -> BigInt {
+        i.as_ref().clone()
+    }
+    pub fn bigint_to_dafny_int(i: &BigInt) -> Rc<BigInt> {
+        Rc::new(i.clone())
+    }
+
+    pub fn dafny_sequence_to_vec<T, X>(s: &Sequence<T>, elem_converter: fn(&T) -> X) -> Vec<X>
+      where T: Clone
+    {
+        let mut array: Vec<T> = Vec::with_capacity(s.cardinality());
+        Sequence::<T>::append_recursive(&mut array, s);
+        array.iter().map(|x| elem_converter(x)).collect()
+    }
+
+    // Used for external conversions
+    pub fn vec_to_dafny_sequence<T, X>(array: &Vec<X>, elem_converter: fn(&X) -> T) -> Rc<Sequence<T>>
+      where T: Clone
+    {
+        let mut result: Vec<T> = Vec::with_capacity(array.len());
+        for elem in array.iter() {
+            result.push(elem_converter(elem));
+        }
+        Sequence::<T>::new_array_sequence(&Rc::new(result))
+    }
+    
+    fn dafny_map_to_hashmap<K, V, K2, V2>(m: &Rc<Map<K, V>>, converter_k: fn(&K)->K2, converter_v: fn(&V)->V2) -> HashMap<K2, V2>
+      where
+          K: Eq + Clone + Hash, V: Clone,
+          K2: Eq + Hash, V2: Clone
+    {
+        m.to_hashmap_owned(converter_k, converter_v)
+    }
+
+    fn hashmap_to_dafny_map<K2, V2, K, V>(map: &HashMap<K2, V2>, converter_k: fn(&K2)->K, converter_v: fn(&V2)->V)
+        -> Rc<Map<K, V>>
+      where
+        K: Eq + Clone + Hash, V: Clone,
+        K2: Eq + Hash, V2: Clone
+    {
+        let mut result: Vec<(K, V)> = Vec::new();
+        for (k, v) in map.iter() {
+            result.push((converter_k(k), converter_v(v)));
+        }
+        let s = Sequence::<(K, V)>::new_array_sequence(&Rc::new(result));
+        Rc::new(Map {
+            data: s,
+            cache: RefCell::new(None)
+        })
+    }
+
+    // --unicode-chars:true
+    pub mod unicode_chars_true {
+        use crate::Sequence;
+        use std::rc::Rc;
+
+        pub fn string_to_dafny_string(s: &str) -> Rc<Sequence<char>> {
+            Sequence::new_array_sequence_is_string(&Rc::new(s.chars().collect()), true)
+        }
+        pub fn dafny_string_to_string(s: &Rc<Sequence<char>>) -> String {
+            let characters = s.to_array();
+            characters.iter().collect::<String>()
+        }
+    }
+    
+    // --unicode-chars:false
+    pub mod unicode_chars_false {
+        use crate::Sequence;
+        use std::rc::Rc;
+        pub fn string_to_dafny_string(s: &str) -> Rc<Sequence<u16>> {
+            Sequence::new_array_sequence_is_string(&Rc::new(s.encode_utf16().collect()), false)
+        }
+        pub fn dafny_string_to_string(s: &Rc<Sequence<u16>>) -> String {
+            let characters = s.to_array();
+            String::from_utf16_lossy(&characters)
+        }
+    }
+    
+}
+
+#[allow(dead_code)]
+enum Sequence<T>
+  where T: Clone,
+{
+    ArraySequence {
+        is_string: bool,
+        node_count: usize,
```

Comment:
Is the node_count actually used by anything at runtime? IIRC it's only in dafnyRuntime.dfy for the benefit of termination proofs in Dafny